sdk/core/azure-core/inc/http/http.hpp

```diff
@@ -63,6 +66,191 @@ enum class HttpMethod
   PATCH,
 };
 
+class Request
+{
+
+private:
+  // query needs to be first or at least before url, since url might update it
+  std::map<std::string, std::string> _queryParameters;
+
+  HttpMethod _method;
+  std::string _url;
+  std::map<std::string, std::string> _headers;
+  std::map<std::string, std::string> _retryHeaders;
+  std::map<std::string, std::string> _retryQueryParameters;
+  // Request can contain no body, or either of next bodies (_bodyBuffer plus size or bodyStream)
+  BodyStream* _bodyStream;
```

Comment:
I noticed you're checking the same code in two different PRs, so I'll just copy-paste my comments from that PR to here.

I understand the request body can be a buffer or a stream, and you want to handle both cases. But I really don't think this is a good idea to do like this. Because we're maintaining two sets of logic all the way from the outmost API into innermost code.

My proposal is storage team wraps the buffer into a stream, and then we only need to deal with stream. We do so in cpplite sdk and have the wrapper off-the-shelf, it works like a charm. See it [here](https://github.com/Azure/azure-storage-cpplite/blob/master/include/mstream.h).

ditto for response.

Reply:
yeah, this is not yet the final ans absolute decision for this.
stream and buffer are currently @JeffreyRichter  desing and it is still under review with all team.
consider this as a not completed work and something that will be definitively be changed

The other PR you might saw is from @gilbertw , he is adding the context to the request and the pipeline.  So at somepoint we will be merging this changes and his.